File: tutorials/007_session_patterns/session_patterns.cpp
// Three Session composition patterns: direct nodes, model.session(), attached session.
//
// Usage:
//   tutorial_v2_007_session_patterns [--mpk /path/to/model.tar.gz]

#include "neat.h"

#include <opencv2/core.hpp>

#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

namespace fs = std::filesystem;

namespace {

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

} // namespace

int main(int argc, char** argv) {
  try {
    const int width = 224;
    const int height = 224;

    cv::Mat rgb(height, width, CV_8UC3, cv::Scalar(80, 40, 160));
    if (!rgb.isContinuous())
      rgb = rgb.clone();

    // CORE LOGIC
    // Pattern 1: build a Session by adding Input/Output nodes directly.
    simaai::neat::Session direct;
    simaai::neat::InputOptions in;
    in.format = "RGB";
    in.width = width;
    in.height = height;
    in.depth = 3;
    in.do_timestamp = true;
    direct.add(simaai::neat::nodes::Input(in));
    direct.add(simaai::neat::nodes::Output());

    std::string mpk;
    if (get_arg(argc, argv, "--mpk", mpk) && fs::exists(mpk)) {
      simaai::neat::Model model(mpk);

      // Pattern 2: ingest the model's default session group.
      simaai::neat::Session from_model;
      from_model.add(model.session());
      std::cout << "model_session_size=" << model.session().size() << "\n";

      // Pattern 3: attach the model under an upstream name with custom options.
      simaai::neat::Model::SessionOptions sopt;
      sopt.include_appsrc = false;
      sopt.include_appsink = true;
      sopt.upstream_name = "camera0";
      sopt.name_suffix = "_camera0";
      sopt.buffer_name = "camera0";

      simaai::neat::Session attached;
      attached.add(model.session(sopt));
      std::cout << "attached_session_size=" << attached.size() << "\n";
    }

    auto run = direct.build(rgb, simaai::neat::RunMode::Sync);
    auto out = run.push_and_pull(rgb, /*timeout_ms=*/1000);
    // END CORE LOGIC

    if (!out.tensor.has_value())
      throw std::runtime_error("direct session output missing tensor");
    std::cout << "direct_rank=" << out.tensor->shape.size() << "\n";
    std::cout << "[OK] 007_session_patterns\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/007_session_patterns/session_patterns.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import os
import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either NEAT is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate\n"
      "If the venv does not exist yet, follow the installation guide."
  )


# CORE LOGIC
def make_base_session(width: int, height: int):
  inp = pyneat.InputOptions()
  inp.format = "RGB"
  inp.width = width
  inp.height = height
  inp.depth = 3
  inp.do_timestamp = True

  s = pyneat.Session()
  s.add(pyneat.nodes.input(inp))
  s.add(pyneat.nodes.output())
  return s
# END CORE LOGIC


def main(argv: list[str]) -> int:
  import numpy as np

  if "--help" in argv:
    print(f"Usage: {argv[0]} [--mpk <path>] [--print-gst]")
    return 0

  strict_mode = os.getenv("SIMA_RUN_TUTORIALS_FULL") is not None

  # Why: runtime markers make intent explicit without requiring external docs.
  # Why: parity/scorecard tooling relies on stable, machine-parseable checkpoints.
  print("STEP input_contract: parse flags and establish deterministic defaults")
  print("STEP run_mode_choice: exercise the chapter's primary runtime path")
  print("STEP output_contract: emit checks and machine-parseable signature")
  print("CHECK strict_flag_available: PASS (strict-mode guard is observable)")
  assert isinstance(strict_mode, bool), "check failed: strict_flag_available (strict-mode guard is observable)"

  width, height = 224, 224
  direct = make_base_session(width, height)

  root = Path(__file__).resolve().parents[2]
  mpk_arg = next((argv[i + 1] for i in range(1, len(argv) - 1) if argv[i] == "--mpk"), None)
  if mpk_arg:
    mpk = Path(mpk_arg)
  else:
    yolo = next(
        (p for p in [
            root / "tmp" / "yolo_v8s_mpk.tar.gz",
            root / "tmp" / "yolov8s_mpk.tar.gz",
            root / "tmp" / "yolo_mpk.tar.gz",
        ] if p.exists()),
        None,
    )
    resnet = next(
        (p for p in [
            root / "tmp" / "resnet_50_mpk.tar.gz",
            root / "tmp" / "resnet50_mpk.tar.gz",
        ] if p.exists()),
        None,
    )
    mpk = next((p for p in [yolo, resnet] if p is not None and p.exists()), None)

  if mpk and mpk.exists():
    # CORE LOGIC
    model = pyneat.Model(str(mpk))
    print(f"model.session().size: {model.session().size()}")

    sopt = pyneat.ModelSessionOptions()
    sopt.include_appsrc = False
    sopt.include_appsink = True
    sopt.upstream_name = "camera0"
    sopt.name_suffix = "_camera0"
    sopt.buffer_name = "camera0"

    attached = pyneat.Session()
    attached.add(model.session(sopt))

    if "--print-gst" in argv:
      print("[direct]")
      print(direct.describe_backend())
      print("[attached]")
      print(attached.describe_backend())
      return 0
    # END CORE LOGIC
  elif "--print-gst" in argv:
    print(direct.describe_backend())
    return 0

  rgb = np.full((height, width, 3), 55, dtype=np.uint8)
  t = pyneat.Tensor.from_numpy(rgb, copy=True, image_format=pyneat.PixelFormat.RGB)
  # CORE LOGIC
  run = direct.build(t, pyneat.RunMode.Sync)
  out = run.run(t, timeout_ms=1000)
  # END CORE LOGIC
  assert out.tensor is not None, "direct session output missing tensor"

  print("CHECK tutorial_completed: PASS (main path reached end without exception)")
  print("SIGNATURE " + json.dumps({
          "tutorial": "007",
          "lang": "py",
          "flow": "chapter_path",
          "run_mode": "sync_or_async",
          "output_kind": "sample_or_tensor",
          "tensor_rank": -1,
          "field_count": -1,
      },
      sort_keys=True,
      separators=(",", ":"),
  ))

  print("[OK] 007_session_patterns")
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
